### PR TITLE
Raise an error if DataArray name is not able to be written to netCDF.

### LINKED ISSUE #533: DataArray.name should always be a string
Consider the following example:

```Python
import numpy as np
import xray

da = xray.DataArray(np.random.random((4, 5)))
ds = da.to_dataset(name=0) # or name=True, or name=(4)
ds.to_netcdf('test.nc')
```
raises this error:
```python
/Users/jhamman/anaconda/lib/python3.4/site-packages/xray/backends/netCDF4_.py in prepare_variable(self, name, variable)
    228             endian='native',
    229             least_significant_digit=encoding.get('least_significant_digit'),
--> 230             fill_value=fill_value)
    231         nc4_var.set_auto_maskandscale(False)
    232 

netCDF4/_netCDF4.pyx in netCDF4._netCDF4.Dataset.createVariable (netCDF4/_netCDF4.c:13217)()

/Users/jhamman/anaconda/lib/python3.4/posixpath.py in normpath(path)
    330     if path == empty:
    331         return dot
--> 332     initial_slashes = path.startswith(sep)
    333     # POSIX allows one or two initial slashes, but treats three or more
    334     # as single slash.

AttributeError: 'int' object has no attribute 'startswith'
```

I think one way to solve this is to cast the name attribute to a string at the time of assignment.  Another way is just to raise an error if a not string variable name is used.

### PR DESCRIPTION
**Pull request by jhamman**
closes #533
xref #549, #551

### CODE REVIEW

File: xray/backends/api.py [hunk 1]
```diff
@@ -62,6 +62,22 @@ def _default_lock(filename, engine):
...
+            if not name:
+                raise ValueError('Invalid name for DataArray or Dataset key: '
+                                 'string must be length 1 or greater')
+        elif name is not None:
+            raise TypeError('DataArray.name or Dataset key must be either a '
+                            'string or None')
+
+    for k, v in dataset.items():
+        check_name(k)
+        check_name(v.name)
```
Comment (shoyer) [inline xray/backends/api.py]: This check is redundant -- `v.name` is always the same as the key.

File: xray/backends/api.py [hunk 2]
```diff
@@ -62,6 +62,22 @@ def _default_lock(filename, engine):
...
+def _validate_dataset_names(dataset):
+    """DataArray.name and Dataset keys must be a string or None"""
+    def check_name(name):
+        if isinstance(name, basestring):
+            if not name:
+                raise ValueError('Invalid name for DataArray or Dataset key: '
+                                 'string must be length 1 or greater')
+        elif name is not None:
+            raise TypeError('DataArray.name or Dataset key must be either a '
+                            'string or None')
```
Comment (shoyer) [inline xray/backends/api.py]: I would add "for serialization to netCDF files"

File: xray/test/test_utils.py [hunk 3]
```diff
@@ -138,3 +138,11 @@ def test_two_numbers(self):
 
     def test_relative_tolerance(self):
         self.assertTrue(utils.is_uniform_spaced([0, 0.97, 2], rtol=0.1))
+
+class Test_hasable(TestCase):
```
Comment (shoyer) [inline xray/test/test_utils.py]: spelling: hasable

File: xray/test/test_backends.py [hunk 4]
```diff
@@ -345,6 +345,17 @@ def test_roundtrip_endian(self):
...
+    def test_invalid_dataarray_names_raise(self):
+        te = (TypeError, 'string or None')
+        ve = (ValueError, 'string must be length 1 or')
+        data = np.random.random((2, 2))
+        da = xray.DataArray(data)
+        for name, e in zip([0, (4, 5), True, ''], [te, te, te, ve]):
+            with self.assertRaisesRegexp(*e):
+                print(name, e)
+                ds = Dataset()
+                ds[name] = da
```
Comment (shoyer) [inline xray/test/test_backends.py]: could just write: `ds = Dataset({name: da})`

File: xray/test/test_backends.py [hunk 5]
```diff
@@ -345,6 +345,17 @@ def test_roundtrip_endian(self):
...
+                ds.to_netcdf('test.nc')
```
Comment (shoyer) [inline xray/test/test_backends.py]: Use `self.roundtrip(ds)` here. Otherwise it only gets tested with one backend.

File: xray/core/dataset.py [hunk 6]
```diff
@@ -612,11 +612,10 @@ def __getitem__(self, key):
         if utils.is_dict_like(key):
             return self.isel(**key)
 
-        key = np.asarray(key)
-        if key.ndim == 0:
-            return DataArray._new_from_dataset(self, key.item())
+        if hashable(key):
```
Comment (shoyer) [inline xray/core/dataset.py]: Add a regression test for this? I think you identified a failing case in your last PR.

Comment (jhamman) [inline xray/core/dataset.py]: done.

**Comment by jhamman** [comment]
@shoyer - can you take a look at these failing tests and see if you can tell which backend is causing the failure?  I'm a bit confused on why it isn't getting caught by my `_validate_dataset_names` function in `to_netcdf`.

**Comment by shoyer** [comment]
The failing test case is on a generic CFEncodedDataStore test, which actually doesn't use `to_netcdf` to roundtrip the data:
https://github.com/xray/xray/blob/c13ee9875b31dceb505b4aead83272fbcbeef27e/xray/test/test_conventions.py#L541

I would simply special case that test on that object because it isn't relevant, e.g., on the `TestCFEncodedDataStore` object, add a method:
```
def test_invalid_dataarray_names_raise(self):
    pass
```

**Comment by jhamman** [comment]
Thanks, that fixed it.  I wasn't aware we had other datastore tests in `test_conventions.py`.  Now I know.  

It looks like I've hit all your previous comments.

**Comment by shoyer** [comment]
Yep, looks good to me. Feel free to merge when you're ready.

**Pull request merged**